File: cpp/205.isomorphic-strings.cpp
class Solution {
public:
  bool isIsomorphic(string s, string t) {
    if (t.size() != s.size()) {
      return false;
    }

    unordered_map<char, char> d1, d2;
    for (int i = 0; i < s.size(); i++) {
      if (d1.find(s[i]) == d1.end() && d2.find(t[i]) == d2.end()) {
        d1[s[i]] = t[i];
        d2[t[i]] = s[i];
      } else if (d1[s[i]] != t[i] || d2[t[i]] != s[i]) {
        return false;
      }
    }

    return true;
  }
};

File: python/205.isomorphic-strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        d1, d2 = {}, {}
        for a, b in zip(s, t):
            if a not in d1 and b not in d2:
                d1[a] = b
                d2[b] = a
            elif a in d1 and d1[a] != b:
                return False
            elif b in d2 and d2[b] != a:
                return False

        return True

